hnn_core/extracellular.py

```diff
@@ -0,0 +1,601 @@
+            H = np.dot(b, a) / norm_a  # NB can be negative
+            L = H + norm_a
+            R2 = np.dot(b, b) - H ** 2  # NB squares
+
+            # To avoid very large values when electrode is placed (anywhere) on
+            # the section axis, enforce minimal perpendicular distance
+            R2 = np.maximum(R2, min_distance ** 2)
+
+            if L < 0 and H < 0:  # electrode is "behind" line segment
+                num = np.sqrt(H ** 2 + R2) - H  # == norm(b) - H
+                denom = np.sqrt(L ** 2 + R2) - L
+            elif L > 0 and H < 0:  # electrode is "on top of" line segment
+                num = (np.sqrt(H ** 2 + R2) - H) * (L + np.sqrt(L ** 2 + R2))
+                denom = R2
+            else:  # electrode is "ahead of" line segment
+                num = np.sqrt(L ** 2 + R2) + L
+                denom = np.sqrt(H ** 2 + R2) + H  # == norm(b) + H
+
+            phi[idx] = np.log(num / denom) / norm_a
+
+    # [dis]: um; [sigma]: S / m
+    # [phi / sigma] = [1/dis] / [sigma] = 1 / [dis] x [sigma]
+    # [dis] x [sigma] = um x (S / m) = 1e-6 S
+    # transmembrane current returned by _ref_i_membrane_ is in [nA]
+    # ==> 1e-9 A x (1 / 1e-6 S) = 1e-3 V = mV
+    # ===> multiply by 1e3 to get uV
+    return 1000.0 * phi / (4.0 * np.pi * sigma)
+
+
+class ExtracellularArray:
+    """Class for recording extracellular potential fields with electrode array
+
+    Note that to add an electrode array to a simulation, you should use the
+    :meth:`hnn_core.Network.add_electrode_array`-method. After simulation,
+    the network will contain a dictionary of `ExtracellularArray`-objects
+    in ``net.rec_array`` (each array must be added with a unique name). An
+    `ExtracellularArray` contains the voltages at each electrode contact,
+    along with the time points at which the voltages were sampled.
+
+    Parameters
+    ----------
+    positions : tuple | list of tuple
+        The (x, y, z) coordinates (in um) of the extracellular electrodes.
+    sigma : float
+        Extracellular conductivity, in S/m, of the assumed infinite,
+        homogeneous volume conductor that the cell and electrode are in.
+    method : str
+        Approximation to use. ``'psa'`` (point source approximation) treats
+        each segment junction as a point extracellular current source.
+        ``'lsa'`` (line source approximation) treats each segment as a line
+        source of current, which extends from the previous to the next segment
+        center point: |---x---|, where x is the current segment flanked by |.
+    min_distance : float (default: 0.5; unit: um)
+        To avoid numerical errors in calculating potentials, apply a minimum
+        distance limit between the electrode contacts and the active neuronal
+        membrane elements that act as sources of current. The default value of
+        0.5 um corresponds to 1 um diameter dendrites.
+    times : None | list of float
```

Comment:
Broadly speaking, what are the `times` and `voltages` arguments for? I see that they are used in tests but not accessible to the user in `Network.add_electrode_array()`.

Reply:
See `__getitem__` in `ExtracellularArray`: a new copy is returned on slicing (which may be surprising, but actually quite functional...)

Reply:
Ah I see. That is pretty slick actually!